libraries/entities/src/index.js

```diff
@@ -0,0 +1,210 @@
+const assert = require('assert').strict;
+const RowClass = require('./RowClass');
+const op = require('./entityops');
+const types = require('./entitytypes');
+const {
+  DUPLICATE_TABLE,
+  NUMERIC_VALUE_OUT_OF_RANGE,
+  UNDEFINED_TABLE,
+  UNIQUE_VIOLATION,
+} = require('taskcluster-lib-postgres');
+
+class Entity {
+  constructor(options) {
+    if (options.context) {
+      assert(options.context instanceof Array, 'context must be an array');
+
+      options.context.forEach(key => {
+        assert(typeof key === 'string', 'elements of options.context must be strings');
+        assert(options.properties[key] === undefined, `property name ${key} is defined in properties and cannot be specified in options.context`);
+        assert(!Reflect.ownKeys(RowClass.prototype).includes(key), `property name ${key} is reserved and cannot be specified in options.context`);
+      });
+    }
+
+    this.partitionKey = options.partitionKey;
+    this.rowKey = options.rowKey;
+    this.properties = options.properties;
+    this.context = options.context || [];
+
+    this.tableName = null;
+    this.db = null;
+    this.serviceName = null;
+    this.contextEntries = null;
+  }
+
+  static op = op;
+
+  static types = types;
+
+  _getContextEntries(context) {
+    const ctx = {};
+
+    assert(typeof context === 'object' && context.constructor === Object, 'context should be an object');
+
+    this.context.forEach(key => {
+      assert(key in context, `context key ${key} must be specified`);
+    });
+
+    Object.entries(context).forEach(([key, value]) => {
+      assert(this.context.includes(key), `context key ${key} was not declared in Entity.configure`);
+      ctx[key] = value;
+    });
+
+    return ctx;
+  }
+
+  _doCondition(conditions) {
```

Comment:
I was thinking a little more about this while walking the dog.  Looking at uses of `load`, `query` and `scan`, I see the following kinds of conditions:

 * things that are used to generate partitionKey / rowKey
 * datestamps (`expires: Entity.op.greaterThanOrEqual(new Date())`)
 * problematic cases
   * `storageType: 'blob'` in `services/queue/src/api.js` -- we can in fact delete this code, as we no longer support blobs
   * `resolveTask` in `services/queue/src/dependencytracker.js` -- we can change this around to just fetch all dependencies of the task (omitting the condition) and check the property with an `if` statement in the JS code.  In this particular case, that won't be especially inefficient.

So, I think we can eliminate the two problematic cases and *only* build conditions for the first two cases -- that is, the only WHERE clause elements we should ever build dynamically are for datestamp properties, where we are given a Date object in JS and can properly convert that to a SQL string without risk of SQL injection.